File: src/main/Python/boj_step16/boj12789.py
import sys
cin = sys.stdin.readline

n = int(cin().strip())
K = 1
line = []
spare = []
peoples = list(map(int, cin().split()))

for person in peoples:
    line.append(person)

while True:
    if not line and not spare:
        print("Nice")
        break

    if not spare:
        if line[0] != K:
            spare.append(line[0])
        else:
            K += 1
        line.pop(0)

    elif not line:
        if spare[-1] != K:
            print("Sad")
            break
        spare.pop()
        K += 1

    else:
        if line[0] != K and spare[-1] != K:
            spare.append(line[0])
            line.pop(0)
        elif line[0] == K:
            line.pop(0)
            K += 1
        elif spare[-1] == K:
            spare.pop()
            K += 1

File: src/main/Cpp/boj_step16/boj12789.cpp
#include <iostream>
#include <stack>
#include <queue>
using namespace std;

int main() {
    queue<int> line;
    stack<int> spare;
    int res = 1;
    int N;
    cin >> N;

    for (int i = 0; i < N; i++) {
        int num;
        cin >> num;
        line.push(num);
    }

    while (true) {
        if (line.empty() && spare.empty()) {
            cout << "Nice";
            break;
        }

        if (spare.empty()) {
            if (line.front() != res)
                spare.push(line.front());
            else
                res++;

            line.pop();
        } else if (line.empty()) {
            if (spare.top() != res) {
                cout << "Sad";
                break;
            }

            spare.pop();
            res++;
        } else {
            if (line.front() != res && spare.top() != res) {
                spare.push(line.front());
                line.pop();
            } else if (line.front() == res) {
                line.pop();
                res++;
            } else if (spare.top() == res) {
                spare.pop();
                res++;
            }
        }
    }
}